Sprint 14

ZakharVP opened the pull request.

Commit:
Сделана задача по спринту 14

Commit:
Добавил в ветку основную страницу

Helvetii951 commented on Tracker/Controllers/10_ChoiceCategoryViewController.swift:
```diff
@@ -0,0 +1,218 @@
+
+import UIKit
+
+class ChoiceCategoryViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
```
Рекомендую не забывать проставлять `final`.

Использование final может помочь компилятору Swift оптимизировать код, так как он знает, что метод или класс не будет переопределен или унаследован. Это может привести к более быстрому выполнению кода, так как компилятор может использовать прямые вызовы методов вместо динамической диспетчеризации.

Про диспетчеризацию, если интересно, можно почитать здесь:
https://habr.com/ru/post/673400/
https://habr.com/ru/company/simbirsoft/blog/673636/

Helvetii951 commented on Tracker/Controllers/10_ChoiceCategoryViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        return cell
+    }
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return TrackerDataStore.shared.getAllCategories().count
+    }
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let categories = TrackerDataStore.shared.getAllCategories()
+        let selectedCategory = categories[indexPath.row].title
+        delegate?.didSelectCategory(selectedCategory)
+        dismiss(animated: true)
+    }
+    
+    private let nameLabel = UILabel()
+    private let imageView = UIImageView()
+    private let doneButton = UIButton()
+    private let tableView = UITableView()
+    private let descriptionLabel = UILabel()
+    
+    override func viewDidLoad() {
```
Код стоит организовывать следующим образом:

Публичные переменные
Приватные переменные
Публичные методы
Приватные методы

Helvetii951 commented on Tracker/Controllers/10_ChoiceCategoryViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        dismiss(animated: true)
+    }
+    
+    private let nameLabel = UILabel()
+    private let imageView = UIImageView()
+    private let doneButton = UIButton()
+    private let tableView = UITableView()
+    private let descriptionLabel = UILabel()
+    
```
Рекомендую всегда использовать [ленивые свойства](https://www.swiftbysundell.com/articles/using-lazy-properties-in-swift/)для объявления подобного.  В нашем случае это какого-то буста не даст. Но, если какая-то вьюха/свойство не будет использоваться, можно сэкономить память. Плюс все необходимые свойства компонента удобно настроить сразу в одном месте.

```swift
private lazy var myAweasomeButton : UIButton = {
  let button = UIButton()
  button.titleLabel?.font = UIFont.systemFont(ofSize: 20, weight: .bold)
  // настройка других свойств 
  // ...
  return button
}() 
```

Подход с ленивыми свойствами используют во многих хороших компаниях.

Helvetii951 commented on Tracker/Controllers/10_ChoiceCategoryViewController.swift:
```diff
@@ -0,0 +1,218 @@
+    
+    private func setupImageView() {
+        imageView.contentMode = .scaleAspectFit
+        imageView.image = UIImage(named: "star_center")
```
Рекомендую генерировать ресурсы(картинки, цвета и т.п). Надежные способы:
1) Использовать библиотеки R.Swift или SwiftGen.
2) Генерить ресурсы с помощью xCode.

В[ этой статье](https://sarunw.com/posts/swift-symbols-for-asset-catalog/) описаны оба.

Helvetii951 commented on Tracker/Controllers/11_NewCategoryViewController.swift:
```diff
@@ -0,0 +1,140 @@
+            doneButton.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -50),
+            doneButton.heightAnchor.constraint(equalToConstant: 60),
+            doneButton.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 20),
+            doneButton.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -20)
+            
+            
+        ])
```
Лучше не делать лишние переносы строк, они могут затруднять чтение кода.

Дополнительный материал про кодстайл:  

- [Подробное описание codestyle](https://google.github.io/swift/) от Google
- [От Ray Wanderlich](https://github.com/raywenderlich/swift-style-guide)
- [На русском от команды tutu](https://github.com/tutu-ru/swift-style-guide)

Helvetii951 commented on Tracker/Controllers/15_ChoiceSheduleViewController.swift:
```diff
@@ -0,0 +1,140 @@
+
+import UIKit
+
+class ChoiceSheduleViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
```
Реализацию `UITableViewDelegate`, `UITableViewDataSource` было бы классно вынести в `extension`. Так код будет чуть более структурирован.

Helvetii951 commented on Tracker/Assets.xcassets/Emoji-4.imageset/Emoji-4.png:
Эмоджи может быть представлено в текстовом представлении. Ты можешь узнать размер шрифта из Figma.

Если коротко, то лучше использовать текст. И память сэкономим.

Helvetii951 commented on Tracker/Controllers/10_ChoiceCategoryViewController.swift:
Нэйминги `10_ChoiceCategoryViewController` и несколько других фалов с префиксом 'n_*` совсем необязательны. Лучше их убрать.

Helvetii951 commented on Tracker/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,241 @@
+    @objc private func showDatePicker() {
+        let datePicker = UIDatePicker()
+        datePicker.datePickerMode = .date
+        datePicker.preferredDatePickerStyle = .wheels
+        datePicker.locale = Locale(identifier: "ru_RU")
+        
```
Не рекомендую привязывать форматтеры к конкретной локали. Пусть текст иногда будет расходиться с дизайном, но пользователь увидит привычное отображение даты.

Helvetii951 commented on Tracker/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,241 @@
+
+import UIKit
+
+class ViewController: UIViewController {
```
`ViewController` не совсем отражает назначение контроллера. Лучше чуть дописать нэйминг. Тебе и коллегам будет понятнее что он делает.

Helvetii951 commented on Tracker/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,241 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Do any additional setup after loading the view.
```
Такие комментарии лучше сразу удалять.

Helvetii951 commented on Tracker/Controllers/ViewController.swift:
```diff
@@ -0,0 +1,241 @@
+        titleLabel.text = "Трекеры"
+        titleLabel.textColor = .black
+        
+        if let sfProBold = UIFont(name: "SFPro-Bold", size: 34) {
```
Не вижу, чтобы был импортирован `SFPro-Bold` в проект. Этот код сразу перескочит на `else`. Лучше просто `UIFont.systemFont(ofSize: 34, weight: .bold)`.

Helvetii951 commented on Tracker/Extention/HabitTrackerController+Ext.swift:
```diff
@@ -0,0 +1,105 @@
+        return cell
+    }
+    
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return 75 // Высота каждой ячейки
+    }
```
Swift - достаточно выразительный язык. Часто можно понять что происходит и без комментариев.

Helvetii951 commented on Tracker/Struct/TrackerCategory.swift:
```diff
@@ -0,0 +1,12 @@
+//
+//  Created by Захар Панченко on 26.04.2025.
+//
+import UIKit
+
+struct TrackerCategory {
+    let title: String
+    var trackers: [Tracker]
+}
```
`UIKit` импортировать совсем не обязательно. Компилятор все оптимизирует, но чем понятнее импорты, тем лучше.

Commit:
Исправил пункты:
Верстка
1. Дизайн заглушки "Что будем отслеживать?".
2. Вертикальные отступы элементов главного экрана (сделал отступы от view).
3. Нет разделителя над таббаром.

Commit:
Исправил следующий пункт
Верстка
4. Дизайн пикера
Добавил локаль ru для отображения дней недели.
Использовал общий компактный стиль.

Commit:
Доработал экран создания нерегулярного события.

Commit:
Функционал
1. Исправил ошибку со скрытием клавиатуры
2. Баг с созданием трекера когда показывались просто категории а не привычки.

Commit:
Поправил остальные замечания

Commit:
Исправил замечания
